### PR TITLE
Resolve compatibility with previous version of C++ client

### PR DESCRIPTION
**Pull request by yush1ga**
### Motivation

The  compatibility with previous versions was lost due to past versions .

### Modifications

I readded some classes and methods for compatibility  with previous versions.

### Result

Previous versions clients can be compiled with latest version's header.

### CODE REVIEW

File: pulsar-client-cpp/include/pulsar/Auth.h [hunk 1]
```diff
@@ -43,33 +61,49 @@ namespace pulsar {
...
-        static AuthenticationPtr Disabled();
+        static AuthenticationPtr getDisabled();
         static AuthenticationPtr create(const std::string& dynamicLibPath);
         static AuthenticationPtr create(const std::string& dynamicLibPath, const std::string& authParamsString);
         static AuthenticationPtr create(const std::string& dynamicLibPath, ParamMap& params);
+        // for compatibility with previous version
+        static AuthDataPtr Disabled();
+
+// for compatibility with previous version
+#ifndef CMS_DISABLE_YCA
```
Comment (saandrews) [inline pulsar-client-cpp/include/pulsar/Auth.h]: PULSAR_DISABLE_YCA ?

**Comment by merlimat** [comment]
I don't think we can or should maintain compatibility with the internal version. One example is that the C++ namespace was already changed (`cms::` to `pulsar::`) , so the source-level compatibility is already gone.

**Comment by saandrews** [comment]
Matteo,

  It's easier to address the namespace difference just by adding
-Dpulsar=cms during compilation. Even dev packages can be generated by
replacing pulsar to cms. This way, we don't have to maintain 2 versions of
headers and made duplicate changes.

Andrews.

**Comment by saandrews** [comment]
👍 

@merlimat your thoughts? I think this is better than maintaining older version of Auth.h and Auth.cc.

**Comment by merlimat** [comment]
@saandrews  I still don't see how source level compatibility can be maintained. Eg: how can you enable YCA? yca auth implementation is not included in this repo

File: pulsar-client-cpp/include/pulsar/Auth.h [hunk 2]
```diff
@@ -26,9 +26,27 @@
 #pragma GCC visibility push(default)
 
 namespace pulsar {
-    
-    typedef std::map<std::string, std::string> ParamMap;
-    
+
+    // for compatibility with previous version
+    enum AuthType {
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Auth.h]: Can we mark the compatibility method as deprecated? This way we can remove them once all the users updated the code.

File: pulsar-client-cpp/include/pulsar/Auth.h [hunk 3]
```diff
@@ -43,33 +61,49 @@ namespace pulsar {
...
-        static AuthenticationPtr Disabled();
+        static AuthenticationPtr getDisabled();
         static AuthenticationPtr create(const std::string& dynamicLibPath);
         static AuthenticationPtr create(const std::string& dynamicLibPath, const std::string& authParamsString);
         static AuthenticationPtr create(const std::string& dynamicLibPath, ParamMap& params);
+        // for compatibility with previous version
+        static AuthDataPtr Disabled();
+
+// for compatibility with previous version
+#ifndef PULSAR_DISABLE_YCA
```
Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Auth.h]: YCA macro logic should be reversed, like `PULSAR_ENABLE_YCA` with default false.

**Comment by merlimat** [comment]
@saandrews @yush1ga Looks good to me. Change a couple of minor things.

File: pulsar-client-cpp/include/pulsar/Auth.h (hunk 2 above)
Comment (yush1ga) [inline pulsar-client-cpp/include/pulsar/Auth.h]: Do you mean like following?
```cpp
#ifdef PULSAR_ENABLE_DEPRECATED_METHOD
enum AuthType {
        AuthNone,
        AuthYcaV1
};
#endif
```

Comment (saandrews) [inline pulsar-client-cpp/include/pulsar/Auth.h]: This good too. In addition, can you also update the comment to say this is deprecated?

Comment (merlimat) [inline pulsar-client-cpp/include/pulsar/Auth.h]: I was thinking about a comment in the code. But the macro looks good

**Pull request merged**